webhooks/networking/ingress_validator.go

```diff
@@ -0,0 +1,129 @@
+		annotationParser:              annotations.NewSuffixAnnotationParser(ingressAnnotationPrefix),
+		classAnnotationMatcher:        ingress.NewDefaultClassAnnotationMatcher(ingConfig.IngressClass),
+		disableIngressClassAnnotation: ingConfig.DisableIngressClassAnnotation,
+		disableIngressGroupAnnotation: ingConfig.DisableIngressGroupNameAnnotation,
+		logger:                        logger,
+	}
+}
+
+var _ webhook.Validator = &ingressValidator{}
+
+type ingressValidator struct {
+	annotationParser              annotations.Parser
+	classAnnotationMatcher        ingress.ClassAnnotationMatcher
+	disableIngressClassAnnotation bool
+	disableIngressGroupAnnotation bool
+	logger                        logr.Logger
+}
+
+func (v *ingressValidator) Prototype(req admission.Request) (runtime.Object, error) {
+	return &networking.Ingress{}, nil
+}
+
+func (v *ingressValidator) ValidateCreate(ctx context.Context, obj runtime.Object) error {
+	ing := obj.(*networking.Ingress)
+	if err := v.checkIngressClassAnnotationUsage(ing, nil); err != nil {
+		return err
+	}
+	if err := v.checkGroupNameAnnotationUsage(ing, nil); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (v *ingressValidator) ValidateUpdate(ctx context.Context, obj runtime.Object, oldObj runtime.Object) error {
+	ing := obj.(*networking.Ingress)
+	oldIng := oldObj.(*networking.Ingress)
+	if err := v.checkIngressClassAnnotationUsage(ing, oldIng); err != nil {
+		return err
+	}
+	if err := v.checkGroupNameAnnotationUsage(ing, oldIng); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (v *ingressValidator) ValidateDelete(ctx context.Context, obj runtime.Object) error {
+	return nil
+}
+
+// checkIngressClassAnnotationUsage checks the usage of kubernetes.io/ingress.class annotation.
+// kubernetes.io/ingress.class annotation cannot be set to the ingress class for this controller once disabled,
+// so that we enforce users to use spec.ingressClassName in Ingress and IngressClass resource instead.
+func (v *ingressValidator) checkIngressClassAnnotationUsage(ing *networking.Ingress, oldIng *networking.Ingress) error {
+	if !v.disableIngressClassAnnotation {
+		return nil
+	}
+	usedInNewIng := false
+	usedInOldIng := false
+	if ingClassAnnotation, exists := ing.Annotations[annotations.IngressClass]; exists {
+		if v.classAnnotationMatcher.Matches(ingClassAnnotation) {
+			usedInNewIng = true
+		}
+	}
+	if oldIng != nil {
+		if ingClassAnnotation, exists := oldIng.Annotations[annotations.IngressClass]; exists {
+			if v.classAnnotationMatcher.Matches(ingClassAnnotation) {
+				usedInOldIng = true
+			}
+		}
+	}
+	if !usedInOldIng && usedInNewIng {
+		return errors.Errorf("new usage of `%s` annotation is forbidden", annotations.IngressClass)
+	}
+	return nil
+}
+
+// checkGroupNameAnnotationUsage checks the usage of "group.name" annotation.
+// "group.name" annotation cannot be set once disabled,
+// so that we enforce users to use spec.group in IngressClass resource instead.
```

Comment:
nit: comment is a bit confusing. IngessClass resource spec doesn't have the spec.group attribute. You might have meant IngressClassParams instead.